# Coursera -- IBM Applied Data Science Capstone Project Week 3


This notebook will be mainly used for IBM course applied data science capstone week 1 project

#### * Please note:The interactive features of the notebook, such as folium maps, will not work in my repository on GitHub. To view my Jupyter notebook with JavaScript content rendered, please [click to view my notebook in nbviewer.](https://nbviewer.jupyter.org/github/J-C-Zh/Coursera-IBM-Data-Science-Professional-9-Applied-Data_Science_Capstone/blob/master/Coursera%20IBM%20Applied%20Data%20Science%20Capstone%20week%203.ipynb)

<h1>Table of contents</h1>

1.Scrape the Toronto postal code dataframe from Wikipedia page. The dataframe should have three columns:'Postal Code','Borough' and       'Neighbourhood'. 
  * Transform the table into a _pandas_ dataframe.
  * Ignore cells with a borough that is Not assigned.
  * For rows that have different neighbourhood names but have the same postal code, combine them into one row with the neighborhoods separated with a comma.
  * If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.

2.Get the latitude and the longitude coordinates of each neighborhood, add 'Latitude' and 'Longitude' columns to the Toronto postal code dataframe.Visualize the data.

3.Explore and cluster the neighborhoods in Toronto. I will work with only boroughs that contain the word 'Toronto '.

## 1. Scrape the Toronto postal code dataframe.

In [1]:
import pandas as pd
from pandas import DataFrame

In [2]:
df_list = pd.read_html('http://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M',header=0)

df_list is a list type data structure which contains all the tables from 'List of postal codes of Canada: M' webpage. We will only need the first table, which is df_list[0]

In [3]:
df = df_list[0]
df.head()

Postal Code          District              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

Change the column names per project requirements.

In [4]:
df.rename(columns={"District": "Borough"}, inplace = True)

Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.

In [5]:
df = df.loc[~(df['Borough'] == 'Not assigned')].reset_index(drop=True)
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
..          ...               ...   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
5              Islington Avenue, Humber Valley Village  
6                                       Malvern, Rouge  
7                                            Don Mills  
8                      Parkview Hill, Woodbine Gardens  
9                             Garden District, Ryerson  
10                                           Glencairn  
11   West Deane Park, Princess Gardens, Martin Grov...  
12              Rouge Hill, Port Union, Highland Creek  
13                                           Don Mills  
14                                    Woodbine Heights  
15                                      St. James Town  
16                                  Humewood-Cedarvale  
17   Eringate, Bloordale Gardens, Old Burnhamthorpe...  
18                   Guildwood, Morningside, West Hill  
19                                         The Beaches  
20                                         Berczy Park  
21                                 Caledonia-Fairbanks  
22                                              Woburn  
23                                             Leaside  
24                                  Central Bay Street  
25                                            Christie  
26                                           Cedarbrae  
27                                   Hillcrest Village  
28     Bathurst Manor, Wilson Heights, Downsview North  

Check if there is more than one neighbourhood existing in one postal code area.

In [6]:
len(df['Postal Code'].unique())

103

There is 103 unique postal codes and the df dataframe has 103 rows. So there is no duplicate postal code existing in our dataframe. Next, let's check if there is 'not assigned' neighbourhood.

In [7]:
df.loc[(df['Neighbourhood'] == 'Not assigned')].shape


(0, 3)

There is no 'not assigned' neighbourhood. The shape of our cleaned dataframe is

In [8]:
df.shape

(103, 3)

In [9]:
print('The dataframe has {} boroughs'.format(
        len(df['Borough'].unique())
    )
)

The dataframe has 10 boroughs


## 2. Get coordinates for each neighbourhood

In order to utilize the Foursquare location data, we need to get the latitude and the longitude coordinates of each neighbourhood.There are two ways to get the geographical coordinates of the neighbourhoods:

1. Using the Geocoder Python package
2. Using the csv file that has the geographical coordinates of each postal code: http://cocl.us/Geospatial_data

I will use both methods to get the data

### 1. Using the Geocoder Python package

In [10]:
!pip install geocoder
print('geocoder installed!')

geocoder installed!


Let's create an empty dataframe called df_cor_geocoder. I will use this new dataframe to store 'Postal Code', 'latitude' and 'longitude' data.

In [11]:
# define the dataframe columns
column_names = ['Postal Code', 'Latitude', 'Longitude'] 

# instantiate the dataframe
df_cor_geocoder = pd.DataFrame(columns=column_names)


I will run a while loop to get cordinates for each postal code.

In [12]:
import geocoder # import geocoder

for code in df['Postal Code']:
    # initialize the variable to None
    lat_lng_coords = None
    # loop until I get the coordinates
    while(lat_lng_coords is None):
        g = geocoder.arcgis('{}, Toronto, Ontario'.format(code))
        lat_lng_coords = g.latlng

    latitude=lat_lng_coords[0]
    longitude=lat_lng_coords[1]
    df_cor_geocoder = df_cor_geocoder.append({'Postal Code': code,
                                          'Latitude': latitude,
                                          'Longitude': longitude}, ignore_index=True)


In [13]:
df_cor_geocoder = df_cor_geocoder.sort_values(by=['Postal Code']).reset_index(drop=True)
df_cor_geocoder.head()

Postal Code  Latitude  Longitude
0         M1B  43.81153  -79.19552
1         M1C  43.78564  -79.15871
2         M1E  43.76575  -79.17520
3         M1G  43.76820  -79.21761
4         M1H  43.76969  -79.23944

In [14]:
df_cor_geocoder.shape

(103, 3)

### 2. Using the csv data
Read the csv data from provided link and store the 'Postal Code', 'Latitude','Longitude' data into a new dataframe called df_cor_csv

In [15]:
df_cor_csv = pd.read_csv('http://cocl.us/Geospatial_data')
df_cor_csv = df_cor_csv.sort_values(by=['Postal Code'])
df_cor_csv.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [16]:
df_cor_csv.shape

(103, 3)

### 3. Compare two cordinate dataframes

df_cor_geocoder and df_cor_csv have the same shape. They both work for getting codinate data. But I can see slightly different result by checking their first five rows. I can also confirm they are different by using _.equals_ function.

In [17]:
df_cor_geocoder.equals(df_cor_csv)

False

The above code returned False, which means the two dataframes have different data. Now I will check how big the difference is. First, Let me create a new dataframe call df_cor_diff. I will caculate the difference of latitude and longtitude data for each postal code and store the difference data in the new dataframe.

In [18]:
df_cor_diff = pd.DataFrame(columns=['Postal Code','Latitude_diff','Longitude_diff'])
df_cor_diff.head()

Empty DataFrame
Columns: [Postal Code, Latitude_diff, Longitude_diff]
Index: []

In [19]:
for i in range(0,df_cor_csv.shape[0]):
    df_cor_diff = df_cor_diff.append({'Postal Code': df_cor_csv['Postal Code'][i],
                                    'Latitude_diff': round((df_cor_geocoder['Latitude'][i]-df_cor_csv['Latitude'][i]),3),
                                    'Longitude_diff': round((df_cor_geocoder['Longitude'][i]-df_cor_csv['Longitude'][i]),3)}, ignore_index=True)

df_cor_diff.head()

Postal Code  Latitude_diff  Longitude_diff
0         M1B          0.005          -0.001
1         M1C          0.001           0.002
2         M1E          0.002           0.014
3         M1G         -0.003          -0.001
4         M1H         -0.003           0.000

In [20]:
print('The biggest latitude difference is {}, and the biggest longitude difference is {}\n'
      .format(abs(df_cor_diff['Latitude_diff'].max()), abs(df_cor_diff['Longitude_diff'].max())))
print('The row with biggest latitude difference:')
print(df_cor_diff.loc[abs(df_cor_diff['Latitude_diff'])==0.012])
print('\n')

print('The row with biggest longitude difference:')
print(df_cor_diff.loc[abs(df_cor_diff['Longitude_diff'])==0.23])
print('\n')

The biggest latitude difference is 0.012, and the biggest longitude difference is 0.23

The row with biggest latitude difference:
   Postal Code  Latitude_diff  Longitude_diff
68         M5V          0.012          -0.005
86         M7R          0.012           0.230


The row with biggest longitude difference:
   Postal Code  Latitude_diff  Longitude_diff
86         M7R          0.012            0.23




The biggest longitude differnce is large compared to latitude difference. More specifically, the two data sets give both biggest  logitude difference and latitude difference for postal code 'M7R'. Let's look more closely. 

In [21]:
print('Data from geocoder method for postal code \'M7R\'')
print(df_cor_geocoder.loc[df_cor_geocoder['Postal Code']=='M7R'])
print('\n')
print('Data from csv file for postal code \'M7R\'')
print(df_cor_csv.loc[df_cor_csv['Postal Code']=='M7R'])

Data from geocoder method for postal code 'M7R'
   Postal Code  Latitude  Longitude
86         M7R  43.64869  -79.38544


Data from csv file for postal code 'M7R'
   Postal Code   Latitude  Longitude
86         M7R  43.636966 -79.615819


By using an [online Latitude/Longtidue Calculator](https://www.nhc.noaa.gov/gccalc.shtml), I know that the distance between (43.64869  -79.38544) and (43.636966 -79.615819) is around 19 km. That is pretty large. For example, if we use geocoder data to calculate the distance between two postal codes, let's say,'M1B' and "M1C', the distance is only about 4 km. However, although the two different cordinate dataset are not 100% same, it won't affect how I visualize and cluster the data. Thus I will go ahead using geocoder dataframe to merge with previous df dataframe to create a new dataframe that has all the information needed for the map. 

In [22]:
# substitute df_cor_geocoder with df_cor_csv if you want to use csv data 
df_cor = df_cor_geocoder

### 4.Data Visualization

Merge df and df_cor as a new dataframe 

In [23]:
df_all = pd.merge(df,df_cor,on ='Postal Code')
df_all

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
..          ...               ...   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  Latitude  Longitude  
0                                            Parkwoods  43.75188  -79.33036  
1                                     Victoria Village  43.73042  -79.31282  
2                            Regent Park, Harbourfront  43.65514  -79.36265  
3                     Lawrence Manor, Lawrence Heights  43.72321  -79.45141  
4          Queen's Park, Ontario Provincial Government  43.66449  -79.39302  
5              Islington Avenue, Humber Valley Village  43.66277  -79.52831  
6                                       Malvern, Rouge  43.81153  -79.19552  
7                                            Don Mills  43.74929  -79.36169  
8                      Parkview Hill, Woodbine Gardens  43.70794  -79.31160  
9                             Garden District, Ryerson  43.65736  -79.37818  
10                                           Glencairn  43.70799  -79.44838  
11   West Deane Park, Princess Gardens, Martin Grov...  43.65279  -79.55406  
12              Rouge Hill, Port Union, Highland Creek  43.78564  -79.15871  
13                                           Don Mills  43.72184  -79.34340  
14                                    Woodbine Heights  43.68970  -79.30680  
15                                      St. James Town  43.65143  -79.37557  
16                                  Humewood-Cedarvale  43.69211  -79.43036  
17   Eringate, Bloordale Gardens, Old Burnhamthorpe...  43.64890  -79.57825  
18                   Guildwood, Morningside, West Hill  43.76575  -79.17520  
19                                         The Beaches  43.67703  -79.29542  
20                                         Berczy Park  43.64531  -79.3

In [24]:
!pip install "folium==0.11.0"
import folium

print('Folium installed and imported!')

Folium installed and imported!


In [25]:
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

In [26]:
address = 'Toronto,Canada'

geolocator = Nominatim(user_agent="Toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [27]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighbourhood in zip(df_all['Latitude'], df_all['Longitude'], df_all['Borough'], df_all['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

## 3.Explore and cluster the neighborhoods in Toronto
Let's simplify the above map and segment and cluster only boroughs that contain the word Toronto.So let's slice the original dataframe and create a new dataframe of the data that has required boroughs name.

In [28]:
name_data = df_all[df_all['Borough'].str.contains('Toronto')].reset_index(drop=True)
name_data.head()

Postal Code           Borough                                Neighbourhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

   Latitude  Longitude  
0  43.65514  -79.36265  
1  43.66449  -79.39302  
2  43.65736  -79.37818  
3  43.65143  -79.37557  
4  43.67703  -79.29542

In [29]:
print('There are {} postal codes for boroughs that contain the word \'Toronto\' in their names.'.format(name_data.shape[0]))

There are 39 postal codes for boroughs that contain the word 'Toronto' in their names.


In [30]:
# create map of Borough that contain the world 'Toronto'
map_name = folium.Map(location=[latitude+0.02, longitude], zoom_start=12)

# add markers to map
for lat, lng, borough, neighbourhood in zip(name_data['Latitude'], name_data['Longitude'], name_data['Borough'], name_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_name)  
    
map_name

Next, we are going to start utilizing the Foursquare API to explore the neighbourhoods and segment them.

In [31]:
#hidden cell codes 
 
CLIENT_ID = '' # your Foursquare ID
CLIENT_SECRET = '' # your Foursquare Secret
VERSION = '' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 
CLIENT_SECRET:


In [32]:
# The code was removed by Watson Studio for sharing.

Let's create a function to get top 100 venues within a radius of 500 meters of all the neighbourhoods in name_data

In [33]:
import requests

radius = 500
LIMIT = 100

def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
# return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Now write the code to run the above function on each neighbourhood and create a new dataframe called *name_venues*.

In [34]:
name_venues = getNearbyVenues(names=name_data['Neighbourhood'],
                                   latitudes=name_data['Latitude'],
                                   longitudes=name_data['Longitude']
                                  )
name_venues.shape


Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West, Forest Hill Road Park
High Park, The Junction South
North Toronto West, Lawrence Park
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
R

(1749, 7)

Let's check the size of the resulting dataframe.

In [35]:
print(name_venues.shape)
name_venues.head()

(1749, 7)


Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0  Regent Park, Harbourfront                43.65514                -79.36265   
1  Regent Park, Harbourfront                43.65514                -79.36265   
2  Regent Park, Harbourfront                43.65514                -79.36265   
3  Regent Park, Harbourfront                43.65514                -79.36265   
4  Regent Park, Harbourfront                43.65514                -79.36265   

                    Venue  Venue Latitude  Venue Longitude  Venue Category  
0        Roselle Desserts       43.653447       -79.362017          Bakery  
1           Tandem Coffee       43.653559       -79.361809     Coffee Shop  
2  Figs Breakfast & Lunch       43.655675       -79.364503  Breakfast Spot  
3         The Yoga Lounge       43.655515       -79.364955     Yoga Studio  
4     Body Blitz Spa East       43.654735       -79.359874             Spa

Let's check how many venues were returned for each neighbourhood

In [36]:
name_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                                
Berczy Park                                                             67   
Brockton, Parkdale Village, Exhibition Place                            87   
Business reply mail Processing Centre, South Ce...                     100   
CN Tower, King and Spadina, Railway Lands, Harb...                      69   
Central Bay Street                                                      71   
Christie                                                                10   
Church and Wellesley                                                    76   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              28   
Davisville North                                                         7   
Dufferin, Dovercourt Village                                            19   
First Canadian Place, Underground city                                 100   
Forest Hill North & West, Forest Hill Road Park                          1   
Garden District, Ryerson                                               100   
Harbourfront East, Union Station, Toronto Islands                       53   
High Park, The Junction South                                            3   
India Bazaar, The Beaches West                                          19   
Kensington Market, Chinatown, Grange Park                               53   
Lawrence Park                                                            2   
Little Portugal, Trinity                                                44   
Moore Park, Summerhill East                                              4   
North Toronto West, Lawrence Park                                        4   
Parkdale, Roncesvalles                                                  46   
Queen's Park, Ontario Provincial Government                             32   
Regent Park, Harbourfront                                               20   
Richmond, Adelaide, King                                               100   
Rosedale                                                                 6   
Roselawn                                                                 1   
Runnymede, Swansea                                                      41   
St. James Town                                                          86   
St. James Town, Cabbagetown                                             44   
Stn A PO Boxes                                                         100   
Studio District                                                         52   
Summerhill West, Rathnelly, South Hill, Forest ...                       6   
The Annex, North Midtown, Yorkville                                     28   
The Beaches                                                              4   
The Danforth West, Riverdale                                             7   
Toronto Dominion Centre, Design Exchange                               100   
University of Toronto, Harbord                                          59   

                                                    Neighbourhood Longitude  \
Neighbourhood                                                                 
Berczy Park                                                              67   
Brockton, Parkdale Village, Exhibition Place                             87   
Business reply mail Processing Centre, South Ce...                      100   
CN Tower, King and Spadina, Railway Lands, Harb...                       69   
Central Bay Street                                                       71   
Christie                                                                 10   
Church and Wellesley                                                     76   
Commerce Court, Victoria Hotel                                          100   
Davisville                                                     

Let's find out how many unique categories can be curated from all the returned venues

In [37]:
print('There are {} uniques categories.'.format(len(name_venues['Venue Category'].unique())))

There are 221 uniques categories.


Analyze each neighbourhood

In [38]:
# one hot encoding
name_onehot = pd.get_dummies(name_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighbourhood column back to datafra
name_onehot['Neighbourhood'] = name_venues['Neighbourhood'] 

# move neighbourhood column to the first column
fixed_columns = [name_onehot.columns[-1]] + list(name_onehot.columns[:-1])
name_onehot = name_onehot[fixed_columns]

name_onehot.head()

Neighbourhood  Accessories Store  American Restaurant  \
0  Regent Park, Harbourfront                  0                    0   
1  Regent Park, Harbourfront                  0                    0   
2  Regent Park, Harbourfront                  0                    0   
3  Regent Park, Harbourfront                  0                    0   
4  Regent Park, Harbourfront                  0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  ...  Trail  Train Station  \
0                 0                   0  ...      0              0   
1                 0                   0  ...      0              0   
2                 0                   0  ...      0              0   
3                 0                   0  ...      0              0   
4                 0                   0  ...      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Wine Bar  Wine Shop  Wings Joint  Women's Store  Yoga Studio  
0         0          0            0              0            0  
1         0          0            0              0            0  
2         0          0            0              0            0  
3         0          0            0              0            1  
4         0          0            0              0            0  

[5 rows x 222 columns]

In [39]:
# new dataframe size
name_onehot.shape

(1749, 222)

Next, let's group rows by neighbourhood and by taking the mean of the frequency of occurrence of each category

In [40]:
name_grouped = name_onehot.groupby('Neighbourhood').mean().reset_index()
name_grouped

Neighbourhood  Accessories Store  \
0                                         Berczy Park           0.000000   
1        Brockton, Parkdale Village, Exhibition Place           0.011494   
2   Business reply mail Processing Centre, South C...           0.000000   
3   CN Tower, King and Spadina, Railway Lands, Har...           0.000000   
4                                  Central Bay Street           0.000000   
5                                            Christie           0.000000   
6                                Church and Wellesley           0.000000   
7                      Commerce Court, Victoria Hotel           0.000000   
8                                          Davisville           0.000000   
9                                    Davisville North           0.000000   
10                       Dufferin, Dovercourt Village           0.000000   
11             First Canadian Place, Underground city           0.000000   
12    Forest Hill North & West, Forest Hill Road Park           0.000000   
13                           Garden District, Ryerson           0.000000   
14  Harbourfront East, Union Station, Toronto Islands           0.000000   
15                      High Park, The Junction South           0.000000   
16                     India Bazaar, The Beaches West           0.000000   
17          Kensington Market, Chinatown, Grange Park           0.000000   
18                                      Lawrence Park           0.000000   
19                           Little Portugal, Trinity           0.000000   
20                        Moore Park, Summerhill East           0.000000   
21                  North Toronto West, Lawrence Park           0.000000   
22                             Parkdale, Roncesvalles           0.000000   
23        Queen's Park, Ontario Provincial Government           0.000000   
24                          Regent Park, Harbourfront           0.000000   
25                           Richmond, Adelaide, King           0.000000   
26                                           Rosedale           0.000000   
27                                           Roselawn           0.000000   
28                                 Runnymede, Swansea           0.000000   
29                                     St. James Town           0.000000   
30                        St. James Town, Cabbagetown           0.000000   
31                                     Stn A PO Boxes           0.000000   
32                                    Studio District           0.000000   
33  Summerhill West, Rathnelly, South Hill, Forest...           0.000000   
34                The Annex, North Midtown, Yorkville           0.000000   
35                                        The Beaches           0.000000   
36                       The Danforth West, Riverdale           0.000000   
37           Toronto Dominion Centre, Design Exchange           0.000000   
38                     University of Toronto, Harbord           0.000000   

    American Restaurant  Antique Shop  Aquarium  Art Gallery  Art Museum  \
0              0.000000      0.000000  0.000000     0.014925    0.000000   
1              0.000000      0.000000  0.000000     0.000000    0.000000   
2              0.020000      0.000000  0.000000     0.010000    0.000000   
3              0.000000      0.000000  0.000000     0.000000    0.000000   
4              0.000000      0.000000  0.000000     0.014085    0.014085   
5              0.000000      0.000000  0.000000     0.000000    0.000000   
6              0.013158      0.000000  0.000000     0.000000    0.000000   
7              0.040000      0.000000  0.000000     0.010000    0.000000   
8              0.000000      0.000000  0.000000     0.000000    0.000000   
9              0.142857      0.000000  0.000000     0.000000    0.000000   
10             0.000000      0.000000  0.000000     0.000000    0.000000   
11             0.040000      0.000000  0.000000     0.010000    0.000000   
12         

In [41]:
# the grouped new size
name_grouped.shape

(39, 222)

In [42]:
#Let's print each neighbourhood along with the top 5 most common venuesnum_top_venues = 5
num_top_venues = 5

for hood in name_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = name_grouped[name_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.09
1          Restaurant  0.04
2  Seafood Restaurant  0.03
3              Bakery  0.03
4         Cheese Shop  0.03


----Brockton, Parkdale Village, Exhibition Place----
         venue  freq
0         Café  0.07
1  Coffee Shop  0.06
2          Bar  0.06
3   Restaurant  0.05
4    Nightclub  0.05


----Business reply mail Processing Centre, South Central Letter Processing Plant Toronto----
                venue  freq
0         Coffee Shop  0.10
1               Hotel  0.05
2  Italian Restaurant  0.04
3                 Bar  0.03
4    Asian Restaurant  0.03


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
                venue  freq
0         Coffee Shop  0.07
1  Italian Restaurant  0.06
2                Café  0.04
3                 Bar  0.04
4      Sandwich Place  0.03


----Central Bay Street----
              venue  freq
0       Coffee Shop  0.11
1 

Let's put that into a pandas dataframe.
First, let's write a function to sort the venues in descending order.

In [43]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [44]:
import numpy as np
# print the top 10 venues for each neighbour
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
neighbourhoods_venues_sorted['Neighbourhood'] = name_grouped['Neighbourhood']

for ind in np.arange(name_grouped.shape[0]):
    neighbourhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(name_grouped.iloc[ind, :], num_top_venues)

neighbourhoods_venues_sorted.head()


Neighbourhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...           Coffee Shop   
3  CN Tower, King and Spadina, Railway Lands, Har...           Coffee Shop   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0            Restaurant                Bakery                   Pub   
1           Coffee Shop                   Bar            Restaurant   
2                 Hotel    Italian Restaurant            Restaurant   
3    Italian Restaurant                  Café                   Bar   
4        Clothing Store      Sushi Restaurant        Cosmetics Shop   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Breakfast Spot        Farmers Market              Beer Bar   
1             Nightclub             Gift Shop        Sandwich Place   
2                   Bar      Asian Restaurant    Salon / Barbershop   
3  Gym / Fitness Center             Speakeasy     French Restaurant   
4                 Plaza            Restaurant        Sandwich Place   

  8th Most Common Venue 9th Most Common Venue  10th Most Common Venue  
0          Cocktail Bar                  Café      Seafood Restaurant  
1           Supermarket   Japanese Restaurant  Furniture / Home Store  
2             Gastropub                   Pub                     Gym  
3     Electronics Store                  Park          Sandwich Place  
4                 Hotel       Bubble Tea Shop       Electronics Store

Cluster Neighbourhoods:
Run k-means to cluster the neighbourhood into 5 clusters.

In [45]:
from sklearn.cluster import KMeans

# set number of clusters
kclusters = 5

name_grouped_clustering = name_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(name_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues.

In [46]:
print(neighbourhoods_venues_sorted.shape)
print(name_data.shape)

(39, 11)
(39, 5)


In [47]:
# The code was removed by Watson Studio for sharing.

In [48]:
# add clustering labels
neighbourhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

name_merged = name_data

# merge name_grouped with name_data to add latitude/longitude for each neighbourhood
name_merged = name_merged.join(neighbourhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

# check the last columns!

name_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

   Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.65514  -79.36265               0           Coffee Shop   
1  43.66449  -79.39302               0           Coffee Shop   
2  43.65736  -79.37818               0           Coffee Shop   
3  43.65143  -79.37557               0           Coffee Shop   
4  43.67703  -79.29542               0     Health Food Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0        Breakfast Spot           Yoga Studio       Thai Restaurant   
1        Sandwich Place                  Park                  Café   
2        Clothing Store                  Café        Cosmetics Shop   
3    Seafood Restaurant                  Café          Cocktail Bar   
4                   Pub                 Trail          Neighborhood   

       5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0       Gym / Fitness Center                   Pub            Restaurant   
1         College Auditorium    Chinese Restaurant          Burger Joint   
2  Middle Eastern Restaurant                 Hotel   Japanese Restaurant   
3        American Restaurant        Farmers Market                 Hotel   
4             Discount Store        Farmers Market                  Farm   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0   Distribution Center     Electronics Store            Event Space  
1            Restaurant        Clothing Store           Cocktail Bar  
2       Bubble Tea Shop             Bookstore       Ramen Restaurant  
3    Italian Restaurant            Restaurant              Gastropub  
4    Falafel Restaurant           Event Space   Ethiopian Restaurant

Finally let's visualize final cluster results.

In [49]:
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(name_merged['Latitude'], name_merged['Longitude'], name_merged['Neighbourhood'], name_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

Examine the Clusters.
Now, I can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, you can then assign a name to each cluster. I will leave this exercise to you.

### Cluster 1

In [50]:
name_merged.loc[name_merged['Cluster Labels'] == 0, name_merged.columns[[1] + list(range(5, name_merged.shape[1]))]]

Borough  Cluster Labels        1st Most Common Venue  \
0   Downtown Toronto               0                  Coffee Shop   
1   Downtown Toronto               0                  Coffee Shop   
2   Downtown Toronto               0                  Coffee Shop   
3   Downtown Toronto               0                  Coffee Shop   
4       East Toronto               0            Health Food Store   
5   Downtown Toronto               0                  Coffee Shop   
6   Downtown Toronto               0                  Coffee Shop   
7   Downtown Toronto               0                         Café   
8   Downtown Toronto               0                  Coffee Shop   
9       West Toronto               0                         Park   
10  Downtown Toronto               0                  Coffee Shop   
11      West Toronto               0                  Coffee Shop   
12      East Toronto               0               Ice Cream Shop   
13  Downtown Toronto               0                  Coffee Shop   
14      West Toronto               0                         Café   
15      East Toronto               0                         Park   
16  Downtown Toronto               0                  Coffee Shop   
17      East Toronto               0                        Diner   
20   Central Toronto               0                         Park   
24   Central Toronto               0           Italian Restaurant   
25      West Toronto               0  Eastern European Restaurant   
26   Central Toronto               0                 Dessert Shop   
27  Downtown Toronto               0                         Café   
28      West Toronto               0                         Café   
30  Downtown Toronto               0                         Café   
31   Central Toronto               0           Light Rail Station   
32  Downtown Toronto               0                  Coffee Shop   
34  Downtown Toronto               0                  Coffee Shop   
35  Downtown Toronto               0                  Coffee Shop   
36  Downtown Toronto               0                  Coffee Shop   
37  Downtown Toronto               0                  Coffee Shop   
38      East Toronto               0                  Coffee Shop   

   2nd Most Common Venue   3rd Most Common Venue  \
0         Breakfast Spot             Yoga Studio   
1         Sandwich Place                    Park   
2         Clothing Store                    Café   
3     Seafood Restaurant                    Café   
4                    Pub                   Trail   
5             Restaurant                  Bakery   
6         Clothing Store        Sushi Restaurant   
7          Grocery Store             Coffee Shop   
8                   Café                   Hotel   
9          Grocery Store  Furniture / Home Store   
10                 Hotel                Aquarium   
11          Cocktail Bar        Asian Restaurant   
12          Intersection                Bus Line   
13                 Hotel                    Café   
14           Coffee Shop                     Bar   
15             Pet Store          Ice Cream Shop   
16            Restaurant                   Hotel   
17           Coffee Shop      Italian Restaurant   
20      Department Store     American Restaurant   
24                  Café      Mexican Restaurant   
25           Coffee Shop                  Bakery   
26                  Café                    Park   
27           Coffee Shop              Restaurant   
28           Coffee Shop                  Bakery   
30           Coffee Shop             Gaming Cafe   
31           Coffee Shop            Liquor Store   
32    Italian Restaurant                    Café   
34                 Hotel      Italian Restaurant   
35           Pizza Place                     Pub   
36                  Café                   Hotel   
37   Japanese Restaurant              Restaurant   
38                 Hotel      Italian Restaurant   

            4th Mo

### Cluster 2

In [51]:
name_merged.loc[name_merged['Cluster Labels'] == 1, name_merged.columns[[1] + list(range(5, name_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
22      West Toronto               1     Convenience Store   
23   Central Toronto               1              Gym Pool   
29   Central Toronto               1     Convenience Store   
33  Downtown Toronto               1          Tennis Court   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
22                  Park        Sandwich Place          Costume Shop   
23                  Park                Garden            Playground   
29            Playground                   Gym                  Park   
33                  Park            Playground        Shop & Service   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
22   Distribution Center        Farmers Market                  Farm   
23        Cosmetics Shop        Discount Store                  Farm   
29        Discount Store                  Farm    Falafel Restaurant   
33            Bike Trail            Campground   Distribution Center   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
22    Falafel Restaurant           Event Space   Ethiopian Restaurant  
23    Falafel Restaurant           Event Space   Ethiopian Restaurant  
29           Event Space  Ethiopian Restaurant      Electronics Store  
33                  Farm    Falafel Restaurant            Event Space

### Cluster 3

In [52]:
name_merged.loc[name_merged['Cluster Labels'] == 2, name_merged.columns[[1] + list(range(5, name_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
21  Central Toronto               2                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
21           Yoga Studio   Distribution Center  Fast Food Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
21        Farmers Market                  Farm    Falafel Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
21           Event Space  Ethiopian Restaurant      Electronics Store

### Cluster 4

In [53]:
name_merged.loc[name_merged['Cluster Labels'] == 3, name_merged.columns[[1] + list(range(5, name_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
18  Central Toronto               3      Business Service   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
18                  Park           Yoga Studio               Dog Run   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
18  Fast Food Restaurant        Farmers Market                  Farm   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
18    Falafel Restaurant           Event Space   Ethiopian Restaurant

### Cluster 5

In [54]:
name_merged.loc[name_merged['Cluster Labels'] == 4, name_merged.columns[[1] + list(range(5, name_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
19  Central Toronto               4          Home Service   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
19           Yoga Studio               Dog Run     Fish & Chips Shop   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
19  Fast Food Restaurant        Farmers Market                  Farm   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
19    Falafel Restaurant           Event Space   Ethiopian Restaurant